### PR TITLE
Implement GetRequiredPlugins for Python

### PR DESCRIPTION
**Pull request by justinvp**
This PR implements `GetRequiredPlugins` for Python, which is called by the engine at the beginning of `pulumi preview|up` to determine/install any missing plugins needed by the program.

This is implemented by running `python -m pip list --format json` to determine the installed Pulumi packages (packages with names that start with `pulumi-`). If the `virtualenv` runtime option is set in `Pulumi.yaml`, it runs the `python` inside the virtual environment. Otherwise, it calls `python` in the usual ways, which works with other tools that manage their own virtual environment (such as `poetry run pulumi up`).

The required plugins are determined based on the Pulumi packages. If a package contains an optional `pulumiplugin.json` file, the information in this file is used to determine the associated plugin: the `name`, `version`, and `server`. There's also a `resource` option that must be set to `true` to indicate this package has an associated plugin, or `false` to indicate the package does not have an associated plugin (e.g. I'd imagine we'd set this to `false` for `pulumi-policy`, in the meantime we hardcode this package as a known package that does not have a plugin). This file is similar to information included in `package.json` for Node.js and `version.txt` in .NET NuGet packages (though, `version.txt` does not currently support specifying a `server`). If the `pulumiplugin.json` file doesn't exist, we attempt to determine the plugin name and version based on the package name and version. This works well for all released packages we have today (which don't have a `pulumiplugin.json` yet), but does not work for alpha/beta/rc/dev builds of these packages because the package version does not include enough information to determine the plugin version for these cases. A commit is included with this PR, which adds support for generating the `pulumiplugin.json` file for packages (via a new schema option). I suspect we'll want to roll this out for all our packages, so `GetRequiredPlugins` will work for dev builds.

This PR also implements support for automatically creating the virtual environment and installing dependencies into it at the beginning of `pulumi preview|up` if the `virtualenv` option is set in `Pulumi.yaml` and the specified directory is missing or empty.

I'd love to add more automated testing around this, but may have to do that as a follow-up item (I don't believe we have much automated testing around similar functionality for the other languages).

Fixes https://github.com/pulumi/pulumi/issues/4848
Fixes https://github.com/pulumi/pulumi/issues/5164
Fixes https://github.com/pulumi/pulumi/issues/5593

### CODE REVIEW

File: pkg/backend/httpstate/policypack.go [hunk 1]
```diff
@@ -321,16 +321,16 @@ func completeNodeJSInstall(finalDir string) error {
...
-		}
-		return nil
-	}); err != nil {
+	const venvDir = "venv"
+	if err := python.InstallDependencies(finalDir, venvDir, false /*showOutput*/); err != nil {
 		return err
 	}
 
+	// Save project with venv info.
+	proj.Runtime.SetOption("virtualenv", venvDir)
```
Comment (stack72) [inline pkg/backend/httpstate/policypack.go:330]: now that this is a requirement - does this mean that the user must specify this in their Pulumi.yaml?

Comment (justinvp) [reply pkg/backend/httpstate/policypack.go:330]: It’s not a requirement. There isn’t any user-facing change with this modification.

File: pkg/codegen/python/importer.go [hunk 2]
```diff
@@ -42,6 +42,8 @@ type PackageInfo struct {
 	Compatibility string `json:"compatibility,omitempty"`
 	// Indicates whether the package generates input/output classes.
 	UsesIOClasses bool `json:"usesIOClasses,omitempty"`
+	// Indicates whether the pulumiplugin.json file should be generated.
+	EmitPulumiPluginFile bool `json:"emitPulumiPluginFile,omitempty"`
```
Comment (lukehoban) [inline pkg/codegen/python/importer.go:46]: Any particular reason we don't want to do this by default? (Vs manually updating all providers to include this?)

File: sdk/python/cmd/pulumi-language-python/main.go [hunk 3]
```diff
@@ -144,8 +148,365 @@ func newLanguageHost(exec, engineAddress, tracing, virtualenv string) pulumirpc.
...
+		Severity:  w.severity,
+	}); err != nil {
+		return 0, err
+	}
+	return len(val), nil
+}
+
+// These packages are known not to have any plugins.
+var packagesWithoutPlugins = map[string]struct{}{
+	"pulumi-policy": {},
```
Comment (lukehoban) [inline sdk/python/cmd/pulumi-language-python/main.go:291]: Should we TODO this to undo the hard coding?

File: pkg/codegen/python/gen.go [hunk 4]
```diff
@@ -1215,8 +1216,34 @@ func sanitizePackageDescription(description string) string {
 	return ""
 }
 
+func genPulumiPluginFile(pkg *schema.Package) (string, error) {
+	type pulumiPlugin struct {
```
Comment (lukehoban) [inline pkg/codegen/python/gen.go:1220]: Can we define and version this serialization format in a single place?  I see this is re-defined in the Python language host implementation.

File: sdk/python/cmd/pulumi-language-python/main.go [hunk 5]
```diff
@@ -144,8 +148,365 @@ func newLanguageHost(exec, engineAddress, tracing, virtualenv string) pulumirpc.
...
+		return nil, errors.Wrapf(err, "parsing `python %s` output", strings.Join(args, " "))
+	}
+
+	// Only return Pulumi packages.
+	var pulumiPackages []pythonPackage
+	for _, pkg := range packages {
+		// We're only interested in packages that start with "pulumi-".
+		if !strings.HasPrefix(pkg.Name, "pulumi-") {
+			continue
+		}
```
Comment (lukehoban) [inline sdk/python/cmd/pulumi-language-python/main.go:321]: This is temporary - right?  In principle, we will want to support 3rd party packages with plugins?

**Review by lukehoban — APPROVED** [review]
> This file is similar to information included in package.json for Node.js and version.txt in .NET NuGet packages (though, version.txt does not currently support specifying a server).

Should we be considering this as a standard format to use across these different language ecosystems?

File: sdk/python/cmd/pulumi-language-python/main.go (hunk 5 above)
Comment (justinvp) [reply sdk/python/cmd/pulumi-language-python/main.go:322]: Yes, temporary. Go and .NET have this same limitation:

https://github.com/pulumi/pulumi/blob/002a228000bdcb10cc9851cc8bce09858065058b/sdk/go/pulumi-language-go/main.go#L140

https://github.com/pulumi/pulumi/blob/002a228000bdcb10cc9851cc8bce09858065058b/sdk/dotnet/cmd/pulumi-language-dotnet/main.go#L251-L252

As we work on making it possible/easier to publish 3rd party plugins, we'll need to address this across all the languages.

**Comment by justinvp** [comment]
> Should we be considering this as a standard format to use across these different language ecosystems?

We certainly could consider it.

- Node.js probably doesn't need it as there's already a `pulumi` section in `package.json` where this information lives.
- .NET has `version.txt` which indicates the NuGet package has a plugin. The file contains the plugin version ([and also now the name of the plugin](https://github.com/pulumi/pulumi/pull/5629)). We could consider further extending `version.txt` to include additional information like the plugin `server`, or consider moving .NET to `pulumiplugin.json`.
- Go doesn't currently have anything like this. It does a best effort of determining the plugin information based on the go modules. There isn't currently a way to specify a custom `server` for the plugin or explicit name or version. I suspect we could adopt `pulumiplugin.json` for Go.

**Comment by lukehoban** [comment]
> Node.js probably doesn't need it as there's already a pulumi section in package.json where this information lives.

I guess I wonder if we should at least standardize on the JSON payload that we include inside the package.json, so that we have a consistent data model for this, even if it lives in different places for different package managers.  Probably not a huge deal - but it feels like consistency here will help avoid capability drift between language ecosystems.

File: pkg/codegen/python/importer.go (hunk 2 above)
Comment (justinvp) [reply pkg/codegen/python/importer.go:46]: Ah, I recall now why I had this as an option. We're going to need to update the `Makefile` of each provider to replace `${PLUGIN_VERSION}` in `pulumiplugin.json` with the plugin version, just like we do with `setup.py`.

https://github.com/pulumi/pulumi-aws/blob/ed9cd60d861bc2d930150c272fa00fac2f774518/Makefile#L51

**Comment by justinvp** [comment]
> standardize on the JSON payload that we include inside the package.json

I went ahead and did this. I defined a common `PulumiPluginJSON` struct that's used by both Python and now Node.js.

**Pull request merged**